demos/array-shell-sort/mpc.js

```diff
@@ -0,0 +1,202 @@
+  function generateOffsets(n) {
+    var offsets = [];
+    for (var offset = Math.floor(n/2); offset > 0; offset = Math.floor(offset / 2)) {
+      offsets.push(offset);
+    }
+    return offsets;
+  }
+
+  // compare exchange two regions of length offset each
+  function compareRegions(a, s, t, offset, mates) {
+    for (var count = 0; count < C; count++) { // Do C amount of compare-exchanges
+      for (var i = 0; i < offset; i++) {
+        var n = a.length;
+        var index1 = s+i;
+        var index2 = t+mates[count][i];
+        if (index1 < n && index2 < n) {
+          compareExchange(a, index1, index2);
+        }
+      }
+    }
+  }
+
+  //Main sorting function
+  function randomizedShellSort(a, offsets, permutationsByOffset) {
+    var n = a.length;
+    for (var j = 0; j < offsets.length; j++) {
+      var offset = offsets[j];
+      var mates = permutationsByOffset[j];
+      // do a shaker pass
+      for (var i = 0; i < n - offset; i += offset) { // compare-exchange up
+        compareRegions(a, i, i + offset, offset, mates);
+      }
+      for (i = n - offset; i >= offset; i -= offset) { // compare-exchange down
+        compareRegions(a, i - offset, i, offset, mates);
+      }
+      // do extended brick pass
+      for (i = 0; i < n - 3 * offset; i += offset) { // compare 3 hops up
+        compareRegions(a, i, i + 3 * offset, offset, mates);
+      }
+      for (i = 0; i < n - 2 * offset; i += offset) { // compare 2 hops up
+        compareRegions(a, i, i + 2 * offset, offset, mates);
+      }
+      for (i = 0; i < n; i += 2 * offset) { // compare odd-even regions
+        compareRegions(a, i, i + offset, offset, mates);
+      }
+      for (i = offset; i < n - offset; i += 2 * offset) { // compare even-odd regions
+        compareRegions(a, i, i + offset, offset, mates);
+      }
+    }
+  }
+
+  /*********************************** Main Computation ********************************************/
+  exports.compute = function (input, jiff_instance) {
+    if (jiff_instance == null) {
+      jiff_instance = saved_instance;
+    }
+    var deferred = $.Deferred();
+    var promise = deferred.promise();
+
+    var final_deferred = $.Deferred();
+    var final_promise = final_deferred.promise();
+
+    // Set-up stage (No MPC here)
+    var offsets = [];
+    var permutationsByOffset = [];
+
+    // Party 1 sends message with offsets and permutation values to all parties.
+    if (jiff_instance.id===1) {
+      var n = input.length;
+      offsets = generateOffsets(n);
+      permutationsByOffset = permutations(offsets);
+      var toSend = [offsets, permutationsByOffset];
+      jiff_instance.emit('preprocess', null, JSON.stringify(toSend));
+    }
+
+    // All parties listen for the message with offsets and permutation values, and store the information in it.
+    jiff_instance.listen('preprocess', function (sender_id, message) {
+      var received = JSON.parse(message);
+      deferred.resolve(received);
```

Comment:
no need for the additional deferred promise here, you can immediately put the code in lines 175-198 here (or after the remove_listener in the next line)

Reply:
Ooh you're right! That is much nicer.

Resolved with last push.